newsreader.py

```diff
@@ -25,65 +26,84 @@ def initConnection(self):
+
+        # Update the last stored post id
+        resp = requests.get(self.conparams[0]+"posts.json", cookies=self.token)
+        posts = json.loads(resp.text)['latest_posts']
         with open(self.lfile, 'w') as f:
-            json.dump(self.groups, f)
+            f.write(str(posts[0]['id']))
         self.initialized = True
 
+    def updateAuthToken(self):
+        self.token = None
+        data={"username":self.conparams[3], "password":self.conparams[4]}
+        backoff = 1
+        while True:
+            resp = requests.post(self.conparams[1], data=data, allow_redirects=False)
+            if resp.ok:
+                for cookie in resp.headers['Set-Cookie'].split(";"):
+                    cookie = cookie.strip()
+                    if cookie.startswith("_t="):
+                        self.token = {"_t":cookie[3:]}
+                        self.time =  time.time()
+                        print(f'[*] New Auth Token: {self.token} acquired at {self.time}')
+                        break
+                break
+            else:
+                print(f'[X] Failed AUTH: {resp.status_code}, Retrying in {backoff} seconds')
+                time.sleep(backoff)
+                backoff = min(60, backoff * 2)
+
     def getIdForTopic(self, topic):
         for cat_id, name in self.categories.items():
             if name == topic:
                 return cat_id
         return -1
 
     def validTopic(self, topic):
-        return topic in self.groups
+        return topic in self.categories.values()
 
     def closest(self, topic, topics=None):
         if topics is None:
-            topics = self.groups
+            topics = self.categories
         for t in topics:
             if t.endswith(topic):
                 return t
         return None
 
-    def updateTopic(self, topic, mention_manager):
+    def updatePosts(self, mention_manager):
         if not self.initialized:
             return []
+        # TODO: Find token expiration, 60 secs is a bit low
         if time.time() - self.time > 60.:
-            self.connect()
-        try:
-            (_, _, first, last, _) = self.conn.group(topic)
-        except Exception as e:
-            print(e, datetime.datetime.now())
-            traceback.print_exc()
-            self.connect()
-            return
-        start = max(self.groups[topic] + 1, first)
-        res = []
-        while start <= last:
-            try:
-                artic = self.conn.article(start)[1]
-                raw_msg = b'\r\n'.join(artic.lines)
-                res.append(newsArticle(raw_msg, mention_manager, topic))
-            except Exception as e:
-                print(e, datetime.datetime.now())
-                traceback.print_exc()
-            start += 1
-        self.groups[topic] = last
+            self.updateAuthToken()
+        
+        # Get latest posts
+        resp = requests.get(self.conparams[0]+'posts.json', cookies=self.token)
```

Comment:
again this could timeout or result in a non 200 response.

Reply:
also it looks like this endpoint requires a `before` parameter (https://docs.discourse.org/#tag/Posts/paths/~1posts.json/get)